File: c++/queue.cpp
#include<iostream>
using namespace std;

#define MAX 4

struct queue {
    int items[MAX];
    int front = -1, rear = -1; // this is not the best practice
};
typedef struct queue queue;

bool is_full(queue &q) {
    return q.rear == MAX - 1;
} 

bool empty(queue &q) {
    return q.front == -1;
} 

void enqueue(queue &q, int num) {
    if (!is_full(q)) {
        if(empty(q)) {
            q.front = 0;
        }
        q.items[++q.rear] = num;
        cout<<"Inserted: "<<num<<endl;
        return;
    }
    cout << "Queue is Full can't insert: "<<num<<endl;
}

int dequeue(queue &q) {
    if (!empty(q)) {
        int num = q.items[q.front];
        if (q.front == q.rear) {
            q.front = -1;
            q.rear = -1;
        } else {
            q.front++;
        }
        cout<<"dequed: "<<num<<endl;;
        return num;
    }
    cout<<"Queue is empty"<<endl;
    return -1;
}

void display(queue &q) {
    if (!empty(q)) {
        cout<<"Elemensts of Queue\n";
        for (int i = q.front; i <= q.rear; i++) {
            cout<<q.items[i]<<endl;
        }
    }
}

int main() {
    queue q;
    enqueue(q, 1);
    enqueue(q, 2);
    dequeue(q);
    display(q);
}

File: java/Queue.java
class Queue {
    private int items[], front, rear, size;

    Queue(int len) {
        items = new int[len];
        front = -1;
        rear = -1;
        size = len;
    }

    public Boolean isFull() {
        return rear == size - 1;
    }

    public Boolean isEmpty() {
        return front == -1;
    }

    public void enQueue(int num) {
        if (!isFull()) {
            if(isEmpty())
                front = 0;
            items[++rear] = num;
            System.out.println("Inserted: "+num);
            return;
        }
        System.out.println("Queue is full can't insert: "+num);
    }
    public int deQueue() {
        if (!isEmpty()) {
            int num = items[front];
            if (front == rear) {
                front = -1;
                rear = -1;
            } else {
                front++;
            }
            return num;
        }
        System.out.println("Queue is empty can't pop");
        return -1;
    }

    public void display() {
        if (!isEmpty()) {
            System.out.println("Elements of queueu are:");
            for (int i = front; i <=rear; i++) {
                System.out.println(items[i]);
            }
        }
    }

    public static void main(String args[]) {
        Queue q = new Queue(5);
        q.enQueue(1);
        q.enQueue(2);
        q.deQueue();
        q.display();
    }
}
